File: examples/09-ipc/polyglot/java/src/main/java/io/orionmesh/demo/Sub.java
package io.orionmesh.demo;

import io.nats.client.Connection;
import io.nats.client.Dispatcher;
import io.nats.client.Nats;

import java.util.Map;
import java.util.concurrent.CountDownLatch;

/**
 * Java NATS subscriber — interoperates with the Rust + Python demos.
 *
 * Args (all optional):
 *   --nats-url     nats://127.0.0.1:4222
 *   --subject      orion.demo.ipc
 *   --queue-group  <name>        // when set, joins a queue group (load-balanced)
 *   --label        java
 *
 * Reads ORION_REPLICA_INDEX env var for the default label.
 */
public class Sub {
    public static void main(String[] argv) throws Exception {
        Map<String, String> args = parseArgs(argv);
        String url = args.getOrDefault("--nats-url",
                System.getenv().getOrDefault("NATS_URL", "nats://127.0.0.1:4222"));
        String subject = args.getOrDefault("--subject", "orion.demo.ipc");
        String queueGroup = args.get("--queue-group");
        String label = args.getOrDefault("--label", labelFromEnv());
        String mode = queueGroup != null ? "queue-group '" + queueGroup + "'" : "fan-out (no queue group)";

        System.out.printf("[java-sub:%s] connecting to %s -> %s (%s)%n", label, url, subject, mode);
        try (Connection nc = Nats.connect(url)) {
            System.out.printf("[java-sub:%s] connected%n", label);
            final String finalLabel = label;
            Dispatcher d = nc.createDispatcher(msg -> {
                String body = new String(msg.getData());
                System.out.printf("[java-sub:%s] recv: %s (subject=%s)%n",
                        finalLabel, body, msg.getSubject());
            });
            if (queueGroup != null) {
                d.subscribe(subject, queueGroup);
            } else {
                d.subscribe(subject);
            }
            System.out.printf("[java-sub:%s] subscribed%n", label);
            // Block forever
            new CountDownLatch(1).await();
        }
    }

    private static String labelFromEnv() {
        String idx = System.getenv("ORION_REPLICA_INDEX");
        return idx != null ? "r" + idx : "java";
    }

    private static Map<String, String> parseArgs(String[] argv) {
        var out = new java.util.HashMap<String, String>();
        for (int i = 0; i + 1 < argv.length; i += 2) {
            out.put(argv[i], argv[i + 1]);
        }
        return out;
    }
}

File: examples/09-ipc/polyglot/python/sub.py
#!/usr/bin/env python3
"""Python NATS subscriber — interoperates with the Rust + Java demos.

Two modes:
  default                  fan-out (every subscriber gets every message)
  --queue-group <name>     load-balanced (one subscriber in the group gets each)

Setup:
  python3 -m venv .venv
  . .venv/bin/activate
  pip install nats-py
"""
from __future__ import annotations

import argparse
import asyncio
import os
import sys

try:
    import nats
except ImportError:
    sys.stderr.write("nats-py not installed. Run: pip install nats-py\n")
    sys.exit(1)


async def main():
    p = argparse.ArgumentParser()
    p.add_argument("--nats-url", default=os.environ.get("NATS_URL", "nats://127.0.0.1:4222"))
    p.add_argument("--subject", default="orion.demo.ipc")
    p.add_argument("--queue-group", default=None)
    p.add_argument("--label", default=None)
    args = p.parse_args()

    label = args.label or (
        f"r{os.environ.get('ORION_REPLICA_INDEX')}"
        if os.environ.get("ORION_REPLICA_INDEX")
        else "py"
    )
    mode = f"queue-group '{args.queue_group}'" if args.queue_group else "fan-out (no queue group)"
    print(f"[py-sub:{label}] connecting to {args.nats_url} -> {args.subject} ({mode})", flush=True)
    nc = await nats.connect(args.nats_url)
    print(f"[py-sub:{label}] connected", flush=True)

    async def handler(msg):
        body = msg.data.decode("utf-8", errors="replace")
        print(f"[py-sub:{label}] recv: {body} (subject={msg.subject})", flush=True)

    if args.queue_group:
        await nc.subscribe(args.subject, queue=args.queue_group, cb=handler)
    else:
        await nc.subscribe(args.subject, cb=handler)
    print(f"[py-sub:{label}] subscribed", flush=True)

    # Stay alive until killed
    await asyncio.Event().wait()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass
